File: frontend/src/app/validators/city.validator.ts
import { AbstractControl, ValidationErrors, ValidatorFn } from '@angular/forms';

export const VALID_CITIES = [
  'Visakhapatnam',
  'Vijayawada',
  'Hyderabad',
  'Ananthapur',
  'Khammam'
] as const;

export function cityValidator(): ValidatorFn {
  return (control: AbstractControl): ValidationErrors | null => {
    if (!control.value) return null;
    return VALID_CITIES.includes(control.value)
      ? null
      : { invalidCity: { value: control.value } };
  };
}

File: backend/src/main/java/com/salon/validation/CityValidator.java
package com.salon.validation;

import jakarta.validation.ConstraintValidator;
import jakarta.validation.ConstraintValidatorContext;

import java.util.Set;

public class CityValidator implements ConstraintValidator<ValidCity, String> {

    private static final Set<String> VALID_CITIES = Set.of(
            "Visakhapatnam", "Vijayawada", "Hyderabad", "Ananthapur", "Khammam"
    );

    @Override
    public boolean isValid(String value, ConstraintValidatorContext context) {
        return value != null && VALID_CITIES.contains(value);
    }
}
